File: code/Java/第13章 排序算法/13.2 归并排序/MergeSort.java
import java.util.Arrays;

public class MergeSort {
	public static void main(String[] args) {
		int[] arr = {5, 8, 6, 3, 9, 2, 1, 7};
		mergeSort(arr);
		System.out.println(Arrays.toString(arr));
	}

	public static void binaryInsertionSort(int[] arr, int left, int right) {
		for (int i = left + 1; i <= right; i++) {
			int temp = arr[i];
			int start = left;
			int end = i - 1;
			while (start <= end) {
				int mid = start + (end - start) / 2;
				if (arr[mid] > temp) {
					end = mid - 1;
				} else {
					start = mid + 1;
				}
			}
			int j;
			for (j = i - 1; j > end; j--) {
				arr[j + 1] = arr[j];
			}
			arr[j + 1] = temp;
		}
	}

	public static void merge(int[] arr, int start, int mid, int end, int[] temp) {
	    int i = start;
	    int j = mid + 1;
	    int k = 0;

	    while(i <= mid && j <= end) {
	        if(arr[i] <= arr[j]) {
	            temp[k++] = arr[i++];
	        } else {
	            temp[k++] = arr[j++];
	        }
	    }

	    while(i <= mid) {
	        temp[k++] = arr[i++];
	    }
	    while(j <= end) {
	        temp[k++] = arr[j++];
	    }

	    for(int m = 0; m < k; i++) {
	        arr[start+m] = temp[m];
	    }
	}

	public static void mergeSortWorker(int[] arr, int start, int end, int[] temp) {
		// 列表长度小于10时，采用二分插入排序
		if(end - start <= 10) {
			binaryInsertionSort(arr, start, end);
			return;
		}
		if(start < end) {
	        int mid = start + (end - start) / 2;
			mergeSortWorker(arr, start, mid, temp);
			mergeSortWorker(arr, mid+1, end, temp);
			// 避免不必要的合并
			if(arr[mid] <= arr[mid+1]) {
				return;
			}
	        merge(arr, start, mid, end, temp);
	    }
	}

	public static void mergeSort(int[] arr) {
		int[] temp = new int[arr.length];
		mergeSortWorker(arr, 0, arr.length - 1, temp);
	}
}

File: code/Python/Chapter8/8-6/merge_sort.py
INESRTION_SORT_THRESHOLD = 10

def merge_sort(lst):
    """Merge Sort (original v1.0)"""
    if len(lst) <= 1:
        return lst
    
    mid = len(lst) // 2
    left_half = lst[:mid]
    right_half = lst[mid:]

    merge_sort(left_half)
    merge_sort(right_half)

    i = 0
    j = 0
    k = 0

    while i < len(left_half) and j < len(right_half):
        if left_half[i] < right_half[j]:
            lst[k] = left_half[i]
            i += 1
        else:
            lst[k] = right_half[j]
            j += 1
        k += 1
    
    while i < len(left_half):
        lst[k] = left_half[i]
        i += 1
        k += 1
    
    while j < len(right_half):
        lst[k] = right_half[j]
        j += 1
        k += 1
    
    return lst

# def insertion_sort(lst):
#     """Binary Insertion Sort"""
#     for i in range(1, len(lst)):
#         temp = lst[i]
#         left = 0
#         right = i - 1

#         while left <= right:
#             mid = (left + right) // 2
#             if temp < lst[mid]:
#                 right = mid - 1
#             else:
#                 left = mid + 1

#         for j in range(i - 1, left - 1, -1):
#             lst[j + 1] = lst[j]
#         lst[left] = temp
    
#     return lst

# def merge_sort(lst):
#     """Merge Sort (optimized v2.0)"""
#     if len(lst) <= INESRTION_SORT_THRESHOLD:
#         return insertion_sort(lst)
    
#     mid = len(lst) // 2
#     left_half = lst[:mid]
#     right_half = lst[mid:]

#     left_half = merge_sort(left_half)
#     right_half = merge_sort(right_half)

#     i = 0
#     j = 0
#     k = 0

#     while i < len(left_half) and j < len(right_half):
#         if left_half[i] < right_half[j]:
#             lst[k] = left_half[i]
#             i += 1
#         else:
#             lst[k] = right_half[j]
#             j += 1
#         k += 1
    
#     while i < len(left_half):
#         lst[k] = left_half[i]
#         i += 1
#         k += 1
    
#     while j < len(right_half):
#         lst[k] = right_half[j]
#         j += 1
#         k += 1
    
#     return lst

# def merge_sort(lst):
#     """Merge Sort (optimized v3.0)"""
#     def merge(lst, left, mid, right):
#         n1 = mid - left + 1
#         n2 = right - mid

#         left_half = [0] * n1
#         right_half = [0] * n2

#         for i in range(0, n1):
#             left_half[i] = lst[left + i]
#         for i in range(0, n2):
#             right_half[i] = lst[mid + i + 1]
        
#         i = 0
#         j = 0
#         k = left

#         while i < n1 and j < n2:
#             if left_half[i] < right_half[j]:
#                 lst[k] = left_half[i]
#                 i += 1
#             else:
#                 lst[k] = right_half[j]
#                 j += 1
#             k += 1
        
#         while i < n1:
#             lst[k] = left_half[i]
#             i += 1
#             k += 1
        
#         while j < n2:
#             lst[k] = right_half[j]
#             j += 1
#             k += 1

#     if len(lst) <= INESRTION_SORT_THRESHOLD:
#         return insertion_sort(lst)

#     current_size = 1
#     while current_size < len(lst) - 1:
#         left = 0
#         while left < len(lst) - 1:
#             mid = min(left + current_size - 1, len(lst) - 1)
#             right = min(left + 2 * current_size - 1, len(lst) - 1)
#             merge(lst, left, mid, right)
#             left = right + 1
#         current_size *= 2
    
#     return lst

lst = [39, 27, 43, 3, 9, 82, 10]
merge_sort(lst)
print(lst)